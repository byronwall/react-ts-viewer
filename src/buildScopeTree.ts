import * as fs from "fs";
import * as path from "path";
import * as ts from "typescript";
import { Position, NodeCategory, ScopeNode } from "./types";

// Position helper (already in types.ts, but ts.LineAndCharacter is 0-based for line)
// export interface Position {
//   line: number; // 1-based
//   column: number; // 0-based
// }

function isScopeBoundary(node: ts.Node): boolean {
  return (
    ts.isSourceFile(node) ||
    ts.isModuleBlock(node) ||
    ts.isClassLike(node) ||
    ts.isFunctionLike(node) || // Catches functions, methods, arrow functions, constructors
    ts.isBlock(node) || // Catches block statements {}
    ts.isCatchClause(node) || // catch (e) {}
    ts.isCaseBlock(node) || // case clauses in a switch actually form a block
    // ts.isCaseClause(node) || // case x: individual case, might be too granular
    // ts.isDefaultClause(node) || // default: individual default, might be too granular
    ts.isForStatement(node) || // for (;;) {}
    ts.isForInStatement(node) || // for (x in y) {}
    ts.isForOfStatement(node) || // for (x of y) {}
    ts.isWhileStatement(node) || // while (true) {}
    ts.isDoStatement(node) // do {} while (true)
  );
}

function mapKindToCategory(
  node: ts.Node,
  sourceFile?: ts.SourceFile
): NodeCategory {
  if (ts.isSourceFile(node)) return NodeCategory.Program;
  if (ts.isModuleDeclaration(node)) return NodeCategory.Module;
  if (ts.isClassLike(node)) return NodeCategory.Class; // Catches ClassDeclaration and ClassExpression
  if (ts.isArrowFunction(node)) {
    // Heuristic: arrow function returning JSX ⇒ React component
    if (
      node.body &&
      (ts.isJsxElement(node.body) ||
        ts.isJsxSelfClosingElement(node.body) ||
        (ts.isBlock(node.body) &&
          ts.forEachChild(node.body, (child) =>
            ts.isReturnStatement(child) &&
            child.expression &&
            (ts.isJsxElement(child.expression) ||
              ts.isJsxSelfClosingElement(child.expression))
              ? true
              : undefined
          )))
    ) {
      return NodeCategory.ReactComponent;
    }
    return NodeCategory.ArrowFunction;
  }
  if (
    ts.isFunctionDeclaration(node) ||
    ts.isMethodDeclaration(node) ||
    ts.isConstructorDeclaration(node) ||
    ts.isGetAccessorDeclaration(node) ||
    ts.isSetAccessorDeclaration(node)
  ) {
    // Heuristic: function returning JSX ⇒ React component
    if (
      node.body &&
      ts.forEachChild(node.body, (c) =>
        ts.isReturnStatement(c) &&
        c.expression &&
        (ts.isJsxElement(c.expression) ||
          ts.isJsxSelfClosingElement(c.expression))
          ? true
          : undefined
      )
    ) {
      return NodeCategory.ReactComponent;
    }
    return NodeCategory.Function;
  }
  if (ts.isBlock(node)) return NodeCategory.Block;
  if (ts.isVariableDeclaration(node)) return NodeCategory.Variable; // Note: ts.VariableDeclarationList is the parent for `const a=1,b=2`
  if (
    ts.isIfStatement(node) ||
    ts.isSwitchStatement(node) ||
    ts.isForStatement(node) ||
    ts.isForInStatement(node) ||
    ts.isForOfStatement(node) ||
    ts.isDoStatement(node) ||
    ts.isWhileStatement(node) ||
    ts.isTryStatement(node) || // try {} catch {} finally {}
    ts.isCatchClause(node)
  )
    return NodeCategory.ControlFlow;

  if (ts.isCallExpression(node)) {
    const identNode = ts.isPropertyAccessExpression(node.expression)
      ? node.expression.name
      : node.expression;
    if (ts.isIdentifier(identNode)) {
      const ident = identNode.getText(sourceFile);
      if (/^use[A-Z]/.test(ident)) return NodeCategory.ReactHook;
    }
    return NodeCategory.Call;
  }
  if (
    ts.isJsxElement(node) ||
    ts.isJsxSelfClosingElement(node) ||
    ts.isJsxFragment(node)
  )
    return NodeCategory.JSX;

  return NodeCategory.Other;
}

function deriveLabel(node: ts.Node, sourceFile?: ts.SourceFile): string {
  if (ts.isFunctionLike(node) && node.name && ts.isIdentifier(node.name))
    return node.name.text;
  if (ts.isClassLike(node) && node.name && ts.isIdentifier(node.name))
    return node.name.text;
  if (ts.isVariableDeclaration(node) && ts.isIdentifier(node.name))
    return node.name.text;
  if (ts.isCallExpression(node)) {
    const expressionText =
      node.expression.getText(sourceFile)?.split("\n")[0] || "";
    return expressionText.length > 50
      ? expressionText.substring(0, 47) + "..."
      : expressionText ||
          ts.SyntaxKind[node.expression.kind] ||
          "CallExpression";
  }
  if (ts.isPropertyAccessExpression(node))
    return node.name.getText(sourceFile) || node.name.escapedText.toString();
  if (ts.isIdentifier(node)) return node.text;
  if (ts.isModuleDeclaration(node))
    return node.name.getText(sourceFile) || node.name.text;
  if (ts.isJsxOpeningElement(node) || ts.isJsxSelfClosingElement(node))
    return node.tagName.getText(sourceFile) || "JSXElement";

  const kindName = ts.SyntaxKind[node.kind];
  return kindName || "UnknownNode"; // Ensure a string is always returned
}

function collectMeta(
  node: ts.Node,
  category: NodeCategory,
  sourceFile?: ts.SourceFile
): Record<string, any> | undefined {
  if (category === NodeCategory.ReactHook && ts.isCallExpression(node)) {
    const expression = ts.isPropertyAccessExpression(node.expression)
      ? node.expression.name
      : node.expression;
    return { hookName: expression.getText(sourceFile) };
  }
  if (category === NodeCategory.ReactComponent && ts.isFunctionLike(node)) {
    return {
      props: node.parameters?.map((p) => p.name.getText(sourceFile)) ?? [],
    };
  }
  if (category === NodeCategory.Variable && ts.isVariableDeclaration(node)) {
    return {
      initializer: node.initializer
        ? node.initializer.getText(sourceFile)
        : undefined,
    };
  }
  return undefined;
}

function lineColOfPos(sourceFile: ts.SourceFile, pos: number): Position {
  const lc = sourceFile.getLineAndCharacterOfPosition(pos);
  return { line: lc.line + 1, column: lc.character }; // TS is 0-indexed for line
}

export function buildScopeTree(
  filePath: string,
  fileText: string = fs.readFileSync(filePath, "utf8")
): ScopeNode {
  const sourceFile = ts.createSourceFile(
    path.basename(filePath), // Use basename for the source file name in AST
    fileText,
    ts.ScriptTarget.Latest,
    true,
    ts.ScriptKind.TSX // Assume TSX, adjust if needed
  );

  const rootLocEnd = lineColOfPos(sourceFile, fileText.length);
  const root: ScopeNode = {
    id: filePath, // Use full path for root ID for uniqueness
    kind: sourceFile.kind,
    category: NodeCategory.Program,
    label: path.basename(filePath),
    loc: {
      start: { line: 1, column: 0 },
      end: rootLocEnd,
    },
    source: fileText,
    value: fileText.length, // Could be LOC too: sourceFile.getLineAndCharacterOfPosition(fileText.length).line + 1
    children: [],
  };

  function walk(node: ts.Node, parentNodeInTree: ScopeNode, sf: ts.SourceFile) {
    let currentContainer = parentNodeInTree;
    const category = mapKindToCategory(node, sf);

    // Create a new scope node if it's a scope boundary OR a significant non-boundary element like a variable declaration or call
    if (
      isScopeBoundary(node) ||
      category === NodeCategory.Variable ||
      category === NodeCategory.ReactHook ||
      category === NodeCategory.Call ||
      category === NodeCategory.JSX
    ) {
      const startPos = node.getStart(sf, /*includeJsDoc*/ false); // false to get actual start
      const endPos = node.getEnd();

      // Skip zero-width nodes or nodes outside parent's range (can happen with some synthetic nodes)
      if (endPos === startPos) {
        ts.forEachChild(node, (child) => walk(child, currentContainer, sf));
        return;
      }

      const startLoc = lineColOfPos(sf, startPos);
      const endLoc = lineColOfPos(sf, endPos);

      // Ensure node is within parent's text span.
      // const parentStartOffset = parentNodeInTree.source === fileText ? 0 : sf.text.indexOf(parentNodeInTree.source);
      // if (parentStartOffset === -1 && parentNodeInTree.category !== NodeCategory.Program) { /* handle error or skip */ }
      // const parentEndOffset = parentStartOffset + parentNodeInTree.source.length;
      // if (startPos < parentStartOffset || endPos > parentEndOffset) {
      //   // This node is outside its AST parent's source range, potentially problematic
      //   // For now, we'll allow it but this could be refined
      // }

      const nodeSourceText = fileText.substring(startPos, endPos);
      const nodeId = `${filePath}:${startPos}-${endPos}`; // Unique ID

      const newNode: ScopeNode = {
        id: nodeId,
        kind: node.kind,
        category: category,
        label: deriveLabel(node, sf),
        loc: { start: startLoc, end: endLoc },
        source: nodeSourceText,
        value: nodeSourceText.length, // char length
        meta: collectMeta(node, category, sf),
        children: [],
      };

      // Check for duplicate IDs (can happen with overlapping nodes or incorrect walk logic)
      // This check is more for debugging the walk logic itself.
      // const checkExisting = (nodes: ScopeNode[], id: string): boolean => {
      //     if (nodes.find(n => n.id === id)) return true;
      //     for (const n of nodes) {
      //         if (checkExisting(n.children, id)) return true;
      //     }
      //     return false;
      // };
      // if (checkExisting([root], newNode.id)) {
      //     console.warn("Duplicate node ID:", newNode.id, "Label:", newNode.label);
      // }

      // Attach to the correct parent in the tree
      // This logic ensures nodes are nested under the *current* semantic container.
      parentNodeInTree.children.push(newNode);
      currentContainer = newNode; // This new node is now the container for its children
    }

    ts.forEachChild(node, (child) => walk(child, currentContainer, sf));
  }

  // Start walking from the SourceFile node itself.
  // The root ScopeNode represents the file, its children will be top-level statements/declarations.
  ts.forEachChild(sourceFile, (child) => walk(child, root, sourceFile));

  return root;
}

// Example Usage (for testing - remove or comment out in production extension code):
// if (require.main === module) {
//   const exampleFilePath = path.join(__dirname, '../../src/webview/App.tsx'); // Adjust path as needed
//   if (fs.existsSync(exampleFilePath)) {
//     const tree = buildScopeTree(exampleFilePath);
//     fs.writeFileSync(path.join(__dirname, 'scopeTreeOutput.json'), JSON.stringify(tree, null, 2));
//     console.log('Scope tree generated to scopeTreeOutput.json');
//   } else {
//     console.error('Example file not found:', exampleFilePath);
//   }
// }
