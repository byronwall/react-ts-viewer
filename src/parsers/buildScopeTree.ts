import * as fs from "fs";
import * as path from "path";
import { ScopeNode, BuildScopeTreeOptions } from "../types";
import { buildScopeTreeForMarkdown } from "./md/buildScopeTreeForMarkdown";
import { buildScopeTreeTs } from "./ts/buildScopeTreeTs";

export function buildScopeTree(
  filePath: string,
  fileText: string = fs.readFileSync(filePath, "utf8"),
  options?: BuildScopeTreeOptions
): ScopeNode {
  console.log("####### buildScopeTree EXECUTING - VERSION X #######", filePath); // Add a version number

  const fileExtension = path.extname(filePath).toLowerCase();

  if (fileExtension === ".md" || fileExtension === ".mdx") {
    // TODO: add options back in
    return buildScopeTreeForMarkdown(filePath, fileText);
  } else {
    return buildScopeTreeTs(filePath, fileText, options);
  }
}
