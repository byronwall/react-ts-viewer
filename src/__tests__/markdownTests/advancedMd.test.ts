import { describe, it } from "vitest";
import { runScopeTreeSnapshotTest } from "../buildScopeTree/testUtils";

describe("Advanced Markdown File Test", () => {
  it("should match snapshot for advanced.md", () => {
    runScopeTreeSnapshotTest({
      snapshotIdentifier: "advancedMd",
      filePath: "advanced.md",
      isFixture: true,
    });
  });
});
