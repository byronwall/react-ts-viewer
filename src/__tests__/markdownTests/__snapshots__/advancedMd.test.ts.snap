// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Advanced Markdown File Test > should match snapshot for advanced.md > advancedMd 1`] = `
{
  "category": "Program",
  "children": [
    {
      "category": "MarkdownHeading",
      "children": [
        {
          "category": "MarkdownHeading",
          "children": [
            {
              "category": "MarkdownHeading",
              "children": [
                {
                  "category": "MarkdownList",
                  "children": [],
                  "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:7:1:3",
                  "kind": 0,
                  "label": "Unordered List",
                  "loc": {
                    "end": {
                      "column": 8,
                      "line": 10,
                    },
                    "start": {
                      "column": 0,
                      "line": 7,
                    },
                  },
                  "meta": {},
                  "source": "- Item 1
  - Nested Item 1.1
  - Nested Item 1.2
- Item 2",
                  "value": 57,
                },
              ],
              "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:5:1:2",
              "kind": 0,
              "label": "H3: Unordered List",
              "loc": {
                "end": {
                  "column": 18,
                  "line": 5,
                },
                "start": {
                  "column": 0,
                  "line": 5,
                },
              },
              "meta": {
                "depth": 3,
              },
              "source": "### Unordered List
- Item 1
  - Nested Item 1.1
  - Nested Item 1.2
- Item 2",
              "value": 76,
            },
            {
              "category": "MarkdownHeading",
              "children": [
                {
                  "category": "MarkdownList",
                  "children": [],
                  "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:14:1:5",
                  "kind": 0,
                  "label": "Ordered List",
                  "loc": {
                    "end": {
                      "column": 13,
                      "line": 18,
                    },
                    "start": {
                      "column": 0,
                      "line": 14,
                    },
                  },
                  "meta": {},
                  "source": "1. First item
2. Second item
   1. Sub-item 2.1
   2. Sub-item 2.2
3. Third item",
                  "value": 80,
                },
              ],
              "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:12:1:4",
              "kind": 0,
              "label": "H3: Ordered List",
              "loc": {
                "end": {
                  "column": 16,
                  "line": 12,
                },
                "start": {
                  "column": 0,
                  "line": 12,
                },
              },
              "meta": {
                "depth": 3,
              },
              "source": "### Ordered List
1. First item
2. Second item
   1. Sub-item 2.1
   2. Sub-item 2.2
3. Third item",
              "value": 97,
            },
          ],
          "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:3:1:1",
          "kind": 0,
          "label": "H2: Lists",
          "loc": {
            "end": {
              "column": 8,
              "line": 3,
            },
            "start": {
              "column": 0,
              "line": 3,
            },
          },
          "meta": {
            "depth": 2,
          },
          "source": "## Lists
### Unordered List
- Item 1
  - Nested Item 1.1
  - Nested Item 1.2
- Item 2
### Ordered List
1. First item
2. Second item
   1. Sub-item 2.1
   2. Sub-item 2.2
3. Third item",
          "value": 183,
        },
        {
          "category": "MarkdownHeading",
          "children": [
            {
              "category": "MarkdownBlockquote",
              "children": [
                {
                  "category": "MarkdownParagraph",
                  "children": [],
                  "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:22:3:8",
                  "kind": 0,
                  "label": "This is a blockquote.
It can s...",
                  "loc": {
                    "end": {
                      "column": 29,
                      "line": 23,
                    },
                    "start": {
                      "column": 2,
                      "line": 22,
                    },
                  },
                  "meta": {},
                  "source": "This is a blockquote.
> It can span multiple lines.",
                  "value": 51,
                },
              ],
              "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:22:1:7",
              "kind": 0,
              "label": "Blockquote",
              "loc": {
                "end": {
                  "column": 29,
                  "line": 23,
                },
                "start": {
                  "column": 0,
                  "line": 22,
                },
              },
              "meta": {},
              "source": "> This is a blockquote.
> It can span multiple lines.
This is a blockquote.
> It can span multiple lines.",
              "value": 105,
            },
          ],
          "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:20:1:6",
          "kind": 0,
          "label": "H2: Blockquotes",
          "loc": {
            "end": {
              "column": 14,
              "line": 20,
            },
            "start": {
              "column": 0,
              "line": 20,
            },
          },
          "meta": {
            "depth": 2,
          },
          "source": "## Blockquotes
> This is a blockquote.
> It can span multiple lines.
This is a blockquote.
> It can span multiple lines.",
          "value": 120,
        },
        {
          "category": "MarkdownHeading",
          "children": [
            {
              "category": "MarkdownHeading",
              "children": [
                {
                  "category": "MarkdownParagraph",
                  "children": [],
                  "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:29:1:11",
                  "kind": 0,
                  "label": "This is .",
                  "loc": {
                    "end": {
                      "column": 22,
                      "line": 29,
                    },
                    "start": {
                      "column": 0,
                      "line": 29,
                    },
                  },
                  "meta": {},
                  "source": "This is \`inline code\`.",
                  "value": 22,
                },
              ],
              "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:27:1:10",
              "kind": 0,
              "label": "H3: Inline Code",
              "loc": {
                "end": {
                  "column": 15,
                  "line": 27,
                },
                "start": {
                  "column": 0,
                  "line": 27,
                },
              },
              "meta": {
                "depth": 3,
              },
              "source": "### Inline Code
This is \`inline code\`.",
              "value": 38,
            },
            {
              "category": "MarkdownHeading",
              "children": [
                {
                  "category": "MarkdownCodeBlock",
                  "children": [],
                  "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:33:1:13",
                  "kind": 0,
                  "label": "Code (javascript)",
                  "loc": {
                    "end": {
                      "column": 3,
                      "line": 38,
                    },
                    "start": {
                      "column": 0,
                      "line": 33,
                    },
                  },
                  "meta": {
                    "lang": "javascript",
                  },
                  "source": "\`\`\`javascript
function greet(name) {
  console.log(\`Hello, \${name}!\`);
}
greet("World");
\`\`\`",
                  "value": 92,
                },
              ],
              "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:31:1:12",
              "kind": 0,
              "label": "H3: Fenced Code Block (JavaScript)",
              "loc": {
                "end": {
                  "column": 34,
                  "line": 31,
                },
                "start": {
                  "column": 0,
                  "line": 31,
                },
              },
              "meta": {
                "depth": 3,
              },
              "source": "### Fenced Code Block (JavaScript)
\`\`\`javascript
function greet(name) {
  console.log(\`Hello, \${name}!\`);
}
greet("World");
\`\`\`",
              "value": 127,
            },
            {
              "category": "MarkdownHeading",
              "children": [
                {
                  "category": "MarkdownCodeBlock",
                  "children": [],
                  "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:42:1:15",
                  "kind": 0,
                  "label": "Code (python)",
                  "loc": {
                    "end": {
                      "column": 3,
                      "line": 47,
                    },
                    "start": {
                      "column": 0,
                      "line": 42,
                    },
                  },
                  "meta": {
                    "lang": "python",
                  },
                  "source": "\`\`\`python
def greet(name):
  print(f"Hello, {name}!")

greet("World")
\`\`\`",
                  "value": 73,
                },
              ],
              "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:40:1:14",
              "kind": 0,
              "label": "H3: Fenced Code Block (Python)",
              "loc": {
                "end": {
                  "column": 30,
                  "line": 40,
                },
                "start": {
                  "column": 0,
                  "line": 40,
                },
              },
              "meta": {
                "depth": 3,
              },
              "source": "### Fenced Code Block (Python)
\`\`\`python
def greet(name):
  print(f"Hello, {name}!")

greet("World")
\`\`\`",
              "value": 104,
            },
          ],
          "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:25:1:9",
          "kind": 0,
          "label": "H2: Code Blocks",
          "loc": {
            "end": {
              "column": 14,
              "line": 25,
            },
            "start": {
              "column": 0,
              "line": 25,
            },
          },
          "meta": {
            "depth": 2,
          },
          "source": "## Code Blocks
### Inline Code
This is \`inline code\`.
### Fenced Code Block (JavaScript)
\`\`\`javascript
function greet(name) {
  console.log(\`Hello, \${name}!\`);
}
greet("World");
\`\`\`
### Fenced Code Block (Python)
\`\`\`python
def greet(name):
  print(f"Hello, {name}!")

greet("World")
\`\`\`",
          "value": 286,
        },
        {
          "category": "MarkdownHeading",
          "children": [
            {
              "category": "MarkdownTable",
              "children": [],
              "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:51:1:17",
              "kind": 0,
              "label": "Table",
              "loc": {
                "end": {
                  "column": 46,
                  "line": 54,
                },
                "start": {
                  "column": 0,
                  "line": 51,
                },
              },
              "meta": {},
              "source": "| Header 1     | Header 2     | Header 3     |
| ------------ | ------------ | ------------ |
| Row 1, Col 1 | Row 1, Col 2 | Row 1, Col 3 |
| Row 2, Col 1 | Row 2, Col 2 | Row 2, Col 3 |",
              "value": 187,
            },
          ],
          "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:49:1:16",
          "kind": 0,
          "label": "H2: Tables",
          "loc": {
            "end": {
              "column": 9,
              "line": 49,
            },
            "start": {
              "column": 0,
              "line": 49,
            },
          },
          "meta": {
            "depth": 2,
          },
          "source": "## Tables
| Header 1     | Header 2     | Header 3     |
| ------------ | ------------ | ------------ |
| Row 1, Col 1 | Row 1, Col 2 | Row 1, Col 3 |
| Row 2, Col 1 | Row 2, Col 2 | Row 2, Col 3 |",
          "value": 197,
        },
        {
          "category": "MarkdownHeading",
          "children": [
            {
              "category": "MarkdownThematicBreak",
              "children": [],
              "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:58:1:19",
              "kind": 0,
              "label": "---",
              "loc": {
                "end": {
                  "column": 3,
                  "line": 58,
                },
                "start": {
                  "column": 0,
                  "line": 58,
                },
              },
              "meta": {},
              "source": "---",
              "value": 3,
            },
          ],
          "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:56:1:18",
          "kind": 0,
          "label": "H2: Horizontal Rule",
          "loc": {
            "end": {
              "column": 18,
              "line": 56,
            },
            "start": {
              "column": 0,
              "line": 56,
            },
          },
          "meta": {
            "depth": 2,
          },
          "source": "## Horizontal Rule
---",
          "value": 22,
        },
        {
          "category": "MarkdownHeading",
          "children": [
            {
              "category": "MarkdownParagraph",
              "children": [
                {
                  "category": "MarkdownImage",
                  "children": [],
                  "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:63:1:22",
                  "kind": 0,
                  "label": "Alt text for image",
                  "loc": {
                    "end": {
                      "column": 54,
                      "line": 63,
                    },
                    "start": {
                      "column": 0,
                      "line": 63,
                    },
                  },
                  "meta": {},
                  "source": "![Alt text for image](https://via.placeholder.com/150)",
                  "value": 54,
                },
              ],
              "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:62:1:21",
              "kind": 0,
              "label": "
",
              "loc": {
                "end": {
                  "column": 54,
                  "line": 63,
                },
                "start": {
                  "column": 0,
                  "line": 62,
                },
              },
              "meta": {},
              "source": "[Link to Google](https://www.google.com)
![Alt text for image](https://via.placeholder.com/150)
![Alt text for image](https://via.placeholder.com/150)",
              "value": 150,
            },
          ],
          "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:60:1:20",
          "kind": 0,
          "label": "H2: Links and Images",
          "loc": {
            "end": {
              "column": 19,
              "line": 60,
            },
            "start": {
              "column": 0,
              "line": 60,
            },
          },
          "meta": {
            "depth": 2,
          },
          "source": "## Links and Images
[Link to Google](https://www.google.com)
![Alt text for image](https://via.placeholder.com/150)
![Alt text for image](https://via.placeholder.com/150)",
          "value": 170,
        },
        {
          "category": "MarkdownHeading",
          "children": [
            {
              "category": "MarkdownList",
              "children": [],
              "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:67:1:24",
              "kind": 0,
              "label": "Unordered List",
              "loc": {
                "end": {
                  "column": 21,
                  "line": 68,
                },
                "start": {
                  "column": 0,
                  "line": 67,
                },
              },
              "meta": {},
              "source": "- [x] Completed task
- [ ] Incomplete task",
              "value": 42,
            },
          ],
          "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:65:1:23",
          "kind": 0,
          "label": "H2: Task Lists",
          "loc": {
            "end": {
              "column": 13,
              "line": 65,
            },
            "start": {
              "column": 0,
              "line": 65,
            },
          },
          "meta": {
            "depth": 2,
          },
          "source": "## Task Lists
- [x] Completed task
- [ ] Incomplete task",
          "value": 56,
        },
      ],
      "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md:1:1:0",
      "kind": 0,
      "label": "H1: Advanced Markdown Features",
      "loc": {
        "end": {
          "column": 28,
          "line": 1,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      "meta": {
        "depth": 1,
      },
      "source": "# Advanced Markdown Features
## Lists
### Unordered List
- Item 1
  - Nested Item 1.1
  - Nested Item 1.2
- Item 2
### Ordered List
1. First item
2. Second item
   1. Sub-item 2.1
   2. Sub-item 2.2
3. Third item
## Blockquotes
> This is a blockquote.
> It can span multiple lines.
This is a blockquote.
> It can span multiple lines.
## Code Blocks
### Inline Code
This is \`inline code\`.
### Fenced Code Block (JavaScript)
\`\`\`javascript
function greet(name) {
  console.log(\`Hello, \${name}!\`);
}
greet("World");
\`\`\`
### Fenced Code Block (Python)
\`\`\`python
def greet(name):
  print(f"Hello, {name}!")

greet("World")
\`\`\`
## Tables
| Header 1     | Header 2     | Header 3     |
| ------------ | ------------ | ------------ |
| Row 1, Col 1 | Row 1, Col 2 | Row 1, Col 3 |
| Row 2, Col 1 | Row 2, Col 2 | Row 2, Col 3 |
## Horizontal Rule
---
## Links and Images
[Link to Google](https://www.google.com)
![Alt text for image](https://via.placeholder.com/150)
![Alt text for image](https://via.placeholder.com/150)
## Task Lists
- [x] Completed task
- [ ] Incomplete task",
      "value": 1069,
    },
  ],
  "id": "/Users/byronwall/Projects/react-ts-viewer/src/__tests__/__fixtures__/advanced.md",
  "kind": 0,
  "label": "advanced.md",
  "loc": {
    "end": {
      "column": 0,
      "line": 69,
    },
    "start": {
      "column": 0,
      "line": 1,
    },
  },
  "meta": {
    "depth": 0,
  },
  "source": "# Advanced Markdown Features

## Lists

### Unordered List

- Item 1
  - Nested Item 1.1
  - Nested Item 1.2
- Item 2

### Ordered List

1. First item
2. Second item
   1. Sub-item 2.1
   2. Sub-item 2.2
3. Third item

## Blockquotes

> This is a blockquote.
> It can span multiple lines.

## Code Blocks

### Inline Code

This is \`inline code\`.

### Fenced Code Block (JavaScript)

\`\`\`javascript
function greet(name) {
  console.log(\`Hello, \${name}!\`);
}
greet("World");
\`\`\`

### Fenced Code Block (Python)

\`\`\`python
def greet(name):
  print(f"Hello, {name}!")

greet("World")
\`\`\`

## Tables

| Header 1     | Header 2     | Header 3     |
| ------------ | ------------ | ------------ |
| Row 1, Col 1 | Row 1, Col 2 | Row 1, Col 3 |
| Row 2, Col 1 | Row 2, Col 2 | Row 2, Col 3 |

## Horizontal Rule

---

## Links and Images

[Link to Google](https://www.google.com)
![Alt text for image](https://via.placeholder.com/150)

## Task Lists

- [x] Completed task
- [ ] Incomplete task
",
  "value": 985,
}
`;
